src/chainstate/stacks/index/trie.rs

```diff
@@ -574,83 +559,84 @@ impl Trie {
-        loop {
-            let prev_block_header = match storage.block_walk(1u32 << depth) {
-                Ok(h) => {
-                    h
-                },
-                Err(e) => {
-                    match e {
-                        Error::NotFoundError => {
-                            // out of blocks
-                            break;
-                        },
-                        _ => {
-                            return Err(e);
-                        }
-                    }
-                }
-            };
+        // here is where some mind-bending things begin to happen.
+        //   we want to find the block at a given _height_. but how to do so?
+        //   use the data stored already in the MARF.
+        let cur_block_height = MARF::get_block_height(storage, &cur_block_header, &cur_block_header)
+            .map_err(|e| match e {
+                Error::NotFoundError => Error::CorruptionError(format!("Could not obtain block height for block {}", &cur_block_header)),
+                x => x
+            })?
+            .ok_or_else(|| Error::CorruptionError(format!("Could not obtain block height for block {}", &cur_block_header)))?;
+        let mut log_depth = 0;
+        while log_depth < 32 && (1u32 << log_depth) <= cur_block_height {
+            let prev_block_header = MARF::get_block_at_height(storage, cur_block_height - (1u32 << log_depth), &cur_block_header)?
+                .ok_or_else(|| Error::CorruptionError(format!("Could not obtain block hash at block height {}",
+                                                              cur_block_height - (1u32 << log_depth))))?;
 
             storage.open_block(&prev_block_header, false)?;
             
             let root_ptr = storage.root_trieptr();
-            storage.read_node_hash_bytes(&root_ptr, hash_buf)?;
 
-            trace!("Include root hash {:?} from block {:?} in ancestor #{}", &to_hex(&hash_buf[hash_buf.len() - TRIEHASH_ENCODED_SIZE..hash_buf.len()]), prev_block_header, 1u32 << depth);
+            let ancestor_hash = storage.read_node_hash_bytes(&root_ptr)?;
+            trace!("Include root hash {:?} from block {:?} in ancestor #{}", 
+                   &to_hex(ancestor_hash.as_ref()), prev_block_header, 1u32 << log_depth);
+
+            hash_buf.push(ancestor_hash);
 
-            depth += 1;
-            storage.open_block(&cur_block_header, false)?;
+            log_depth += 1;
+        }
+
+        Ok(hash_buf)
+    }
+
+    /// Calculate the byte vector of the ancestor root hashes of this trie.
+    /// s must point to the block that contains the trie's root.
+    pub fn get_trie_ancestor_hashes_bytes(storage: &mut TrieFileStorage) -> Result<Vec<TrieHash>, Error> {        
+        let cur_block_header = storage.get_cur_block();
+        let cur_block_rw = storage.readwrite();
+
+        if let Some(cached_ancestor_hashes_bytes) = storage.check_cached_ancestor_hashes_bytes(&cur_block_header) {
```

Comment:
Should this `if/then` block be inside the storage impl?  I feel like this is exposes too much of the inner workings.

Reply:
That was my first thought, but right now, the storage interface is such that calls into the storage interface won't ever open a new block themselves, instead relying on the caller to do that. The `inner_get_trie_ancestor_hashes_bytes` function does need to open blocks, so moving the block would violate that property.